File: c/test/other/house_robber.c
/* 

You are planning to rob houses on a specific street, and you know that every house on the street has a certain amount of money hidden. The only thing stopping you from robbing all of them in one night is that adjacent houses on the street have a connected security system. The system will automatically trigger an alarm if two adjacent houses are broken into on the same night.

Given a list of non-negative integers nums representing the amount of money hidden in each house, determine the maximum amount of money you can rob in one night without triggering an alarm.

Example

For nums = [1, 1, 1], the output should be
houseRobber(nums) = 2.

The optimal way to get the most money in one night is to rob the first and the third houses for a total of 2.

 */

#include<stdio.h>
#include<stdlib.h>

#define MAX(a,b) (((a) >= (b)) ? (a) : (b))

static int test[] = { 1, 2, 3, 1 };

static int rob(int *nums, int numSize);

int main()
{
    printf("%d \n", rob(test, 4));
    getchar();
    return 0;
}


static int rob(int* nums, int numsSize){
    if(numsSize == 0)
    {
        return 0;
    }
    
    if(numsSize == 1)
    {
        return nums[0];
    }
    
    if(numsSize == 2)
    {
        return MAX(nums[0], nums[1]);
    }
    
    int mem[2];
    mem[0] = nums[0];
    mem[1] = MAX(nums[0], nums[1]);
    
    int i = 2;
    for(; i < numsSize; ++i)
    {
        int curr = MAX(mem[1], (nums[i] + mem[0]));
        mem[0] = mem[1];
        mem[1] = curr;
    }
    return mem[1] ;
}

File: python/psets/house_robber.py
"""
You are a professional robber planning to rob houses along a street. Each house has a certain amount of money stashed, the only constraint stopping you from robbing each of them is that adjacent houses have security systems connected and it will automatically contact the police if two adjacent houses were broken into on the same night.

Given an integer array nums representing the amount of money of each house, return the maximum amount of money you can rob tonight without alerting the police.
"""

from typing import List


def rob(nums: List[int]) -> int:
    n = len(nums)
    if n == 0: return 0
    if n == 1: return nums[0]
    if n < 3: return max(nums[0], nums[1])
    ans = [0 for _ in nums]
    ans[0] = nums[0]
    ans[1] = max(nums[0], nums[1])

    for i in range(2,n):
        ans[i] = max(nums[i]+ans[i-2], ans[i-1])
    return max(ans)

if __name__ == "__main__":
    print(rob([1,2,3,1]))  # Expected 4
    print(rob([2,7,9,3,1]))# Expected 12